File: bench/cases/array-map-1m/main.ts
function loopSum(n: number, k: number): number {
  const xs: number[] = []
  for (let i = 0; i < n; i = i + 1) {
    xs.push(i)
  }
  const ys: number[] = xs.map((x: number): number => x + k)
  let sum = 0
  for (let i = 0; i < ys.length; i = i + 1) {
    sum = sum + ys[i]
  }
  return sum
}

console.log(loopSum(10000000, 2))

File: bench/cases/array-map-1m/main.rs
use std::hint::black_box;

#[inline(never)]
fn loop_sum(n: i64, k: i64) -> i64 {
    let mut xs: Vec<i64> = Vec::new();
    for i in 0..n {
        xs.push(i);
    }
    // Force the closure to be a real boxed callable so the indirect
    // call cost matches torajs's Closure path. Without black_box +
    // explicit Fn dyn, rustc inlines + vectorizes the loop into pure
    // arithmetic and the comparison is no longer apples-to-apples.
    let f: Box<dyn Fn(i64) -> i64> = Box::new(move |x| x + k);
    let f_ref: &dyn Fn(i64) -> i64 = black_box(&*f);
    let ys: Vec<i64> = xs.iter().map(|&x| f_ref(x)).collect();
    let mut sum: i64 = 0;
    for &y in &ys {
        sum = sum + y;
    }
    sum
}

fn main() {
    println!("{}", loop_sum(10_000_000, 2));
}
